src/RandUtil.chpl

```diff
@@ -0,0 +1,90 @@
+module RandUtil {
+    use MultiTypeSymbolTable;
+    use MultiTypeSymEntry;
+    use RandMsg;
+    use ArkoudaRandomCompat;
+
+    const minPerStream = 256; // minimum number of elements per random stream
+
+    record scalarOrArray {
+        var isArray: bool;
+        var sym;  // TODO figure out type hint here to avoid generic
+        var val: real;
+
+        proc init(scalarOrArrayString: string, isArray: bool, st: borrowed SymTab) {
+            // I'm not sure if there's a good way to remove these try!
+            this.isArray = isArray;
+            if isArray {
+                try! st.checkTable(scalarOrArrayString);
+                this.sym = try! toSymEntry(getGenericTypedArrayEntry(scalarOrArrayString, st),real).a;
+            }
+            else {
+                // prob not the smartest way of doing this
+                // we just want to avoid unnecessarily creating a large array
+                this.sym = try! makeDistArray([0.0]);
+                val = try! scalarOrArrayString:real;
+            }
+        }
+
+        proc this(idx): real {
+            return if isArray then this.sym[idx] else this.val;
+        }
+    }
+
+    enum GenerationFunction {
+      PoissonGenerator,
+    }
+
+    proc uniformStreamPerElem(ref randArr: [?D] ?t, param function: GenerationFunction, hasSeed: bool, const lam: scalarOrArray(?), ref rng) throws {
```

Comment:
If it would make things simpler when you add the other generators, I think you could adjust this s.t. the generator function can be passed in as an argument directly so you don't have to use the enum+select below: 
```chapel
proc uniformStreamPerElem(ref randArr: [?D] ?t, function: proc(_: real, ref _: randomStream(real)), ...
```

(although I'm not sure if all the generator functions would have the same signature?)

Reply:
ah yah I didn't consider this! but yah I don't think they would have the same signature